### PR TITLE
Add upload media functionality

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Reference:
https://github.com/terraform-providers/terraform-provider-vcd/issues/114

Add upload media image functionality
* find media image
* delete media image
* code refactoring
* unit tests
* verified against 8.2, 9.1, 9.5

Note: There is two possibilities upload media - against vdc and catalog. Both features are implemented and using different api to achieve that, cross functionality is reused. When upload media for vdc it's uploaded to automatically created folder "Fresh".

### CODE REVIEW

File: govcd/api.go [hunk 1]
```diff
@@ -27,10 +27,7 @@ type Client struct {
 	Http          http.Client // HttpClient is the client to use. Default will be used if not provided.
 }
 
-// NewRequest creates a new HTTP request and applies necessary auth headers if
-// set.
-func (cli *Client) NewRequest(params map[string]string, method string, reqUrl url.URL, body io.Reader) *http.Request {
-
+func (cli *Client) NewRequestWitNotEncodedParams(params map[string]string, notEncodeParams map[string]string, method string, reqUrl url.URL, body io.Reader) *http.Request {
```
Comment (dataclouder) [inline govcd/api.go]: What's the need for `NewRequestWithNotEndodedParams`? I think I could guess, but a detailed commend on why it is here and when to use it in preference to `NewRequest` should be needed.
If I guess correctly the need for this function, then the argument `notEncodeParams` should be `notEncodedParams`.

File: govcd/api.go [hunk 2]
```diff
@@ -41,6 +38,10 @@ func (cli *Client) NewRequest(params map[string]string, method string, reqUrl ur
 	// Add the params to our URL
 	reqUrl.RawQuery = reqValues.Encode()
 
+	for key, value := range notEncodeParams {
```
Comment (dataclouder) [inline govcd/api.go]: 1. Do we need to check whether `value` is empty?
2. What if `key` is empty?
3. What if `rawQuery` is empty? We'd get an incorrect query then (starting with `&`)

File: govcd/org.go [hunk 3]
```diff
@@ -211,10 +211,18 @@ func (adminOrg *AdminOrg) GetVdcByName(vdcname string) (Vdc, error) {
...
+			if len(splitbyAdminHREF) == 1 {
+				vdcHREF = vdcs.HREF
+			} else {
+				vdcHREF = splitbyAdminHREF[0] + splitbyAdminHREF[1]
+			}
+
 			vdcURL, err := url.ParseRequestURI(vdcHREF)
 			if err != nil {
-				return Vdc{}, fmt.Errorf("Error parsing url: %v", err)
+				return Vdc{}, fmt.Errorf("eror parsing url: %v", err)
```
Comment (dataclouder) [inline govcd/org.go]: 'eror' -> 'error'

File: govcd/api.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/api.go]: Added comment.

File: govcd/api.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/api.go]: Added check to verify value and key, also starting with & is ok. Verified.

File: govcd/api_vcd_test.go [hunk 4]
```diff
@@ -78,6 +80,10 @@ type TestConfig struct {
 		OVAPath        string `yaml:"ovaPath,omitempty"`
 		OVAChunkedPath string `yaml:"ovaChunkedPath,omitempty"`
 	} `yaml:"ova"`
+	Media struct {
+		ISOPath string `yaml:"isoPath,omitempty"`
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/api_vcd_test.go]: Renamed to mediaPath. According VDC can iso or ftl files, at the moment we decided to support only iso, as floppy usage under question.

File: govcd/catalog_test.go [hunk 5]
```diff
@@ -271,3 +269,139 @@ func skipWhenOvaPathMissing(vcd *TestVCD, check *C) {
 		check.Skip("Skipping test because no ova path given")
 	}
 }
+
+// Tests System function UploadMediaImage by checking if provided standard iso file uploaded.
+func (vcd *TestVCD) Test_CatalogUploadMediaImage(check *C) {
+	skipWhenIsoPathMissing(vcd, check)
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/catalog_test.go]: Current implementation supports system admin and org admin

File: govcd/catalog_test.go [hunk 6]
```diff
@@ -271,3 +269,139 @@ func skipWhenOvaPathMissing(vcd *TestVCD, check *C) {
...
+
+	AddToCleanupList(itemName, "mediaImage", vcd.org.Org.Name+"|"+vcd.vdc.Vdc.Name, "Test_UploadCatalogMediaImage")
+
+	mediaItem, err := vcd.vdc.FindMediaImage(itemName)
+	check.Assert(err, IsNil)
+	check.Assert(mediaItem, Not(Equals), MediaItem{})
+
+	task, err := mediaItem.Delete()
+	check.Assert(err, IsNil)
+	task.WaitTaskCompletion()
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/catalog_test.go]: fixed

File: govcd/media.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+		uploadLink:               uploadLink.String(), // just take string
+		uploadedBytes:            0,
+		fileSizeToUpload:         fileSize,
+		uploadPieceSize:          uploadPieceSize,
+		uploadedBytesForCallback: 0,
+		allFilesSize:             fileSize,
+		callBack:                 callBack,
+	}
+
+	go uploadFile(client, mediaFilePath, details)
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media.go]: General ways of handling error isn't good enough due specific handling on showing progress of upload. Added handling of error with callback approach.

File: govcd/media.go [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+				}
+			}
+		}
+	} else {
+		util.Logger.Printf("[Error] Failed to delete media item created with error: %v", media)
+	}
+}
+
+func queryMedia(client *Client, mediaUrl string, newItemName string) (*types.Media, error) {
+	util.Logger.Printf("[TRACE] Qeurying media: %s\n", mediaUrl)
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media.go]: Fixed

File: govcd/media.go [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+	if headerOk {
+		return true, nil
+	} else {
+		return false, errors.New("file header didn't match ISO standard")
+	}
+}
+
+// verify file header info: https://www.garykessler.net/library/file_sigs.html
+func verifyHeader(buf []byte) bool {
+	return (buf[32769] == 0x43 && buf[32770] == 0x44 &&
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/media.go]: Added comments.

File: govcd/api.go [hunk 10]
```diff
@@ -25,12 +25,11 @@ type Client struct {
 	VCDAuthHeader string      // Authorization header
 	VCDHREF       url.URL     // VCD API ENDPOINT
 	Http          http.Client // HttpClient is the client to use. Default will be used if not provided.
+	SysAdmin      bool        // flag if client is connected as system administrator
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply govcd/api.go]: 2) Fixed
1) Is setted when authorize happens, at the moment we do that checking is org is System. Other places has similar checking so now can be used one way.

**Pull request merged**